### PR TITLE
PowerPC C ZST ABI fixes

### PR DESCRIPTION
**Pull request by smaeul**
These patches fix some longstanding issues with the PowerPC C ABI that were causing several test failures. With these patches, I get a clean testsuite run on `powerpc-unknown-linux-musl`.

The [PowerPC SVR4 ELF psABI](http://refspecs.linux-foundation.org/elf/elfspec_ppc.pdf) specifies that, for parameter passing:

> A struct, union, or long double, any of which shall be treated as
a pointer to the object, or to a copy of the object where necessary to
enforce call-by-value semantics. Only if the caller can ascertain that the
object is "constant" can it pass a pointer to the object itself.

The first commit fixes this for most aggregates; the second commit fixes this for zero-size aggregates.

### CODE REVIEW

**Comment by Alexendoo** [comment]
Ping from triage, any updates? @nagisa

**Comment by hdhoang** [comment]
from triaging, requesting a different reviewer. ping @oli-obk can you take a look?

**Comment by nagisa** [comment]
@bors r+

Sorry for taking so long to look at this

File: src/librustc/ty/layout.rs [hunk 1]
```diff
@@ -2648,12 +2648,11 @@ where
...
         let target = &cx.tcx().sess.target.target;
-        let win_x64_gnu =
-            target.target_os == "windows" && target.arch == "x86_64" && target.target_env == "gnu";
-        let linux_s390x =
-            target.target_os == "linux" && target.arch == "s390x" && target.target_env == "gnu";
-        let linux_sparc64 =
-            target.target_os == "linux" && target.arch == "sparc64" && target.target_env == "gnu";
+        let indirect_zst = match target.arch.as_ref() {
+            "powerpc" | "s390x" | "sparc64" => true,
+            "x86_64" => target.target_os == "windows" && target.target_env == "gnu",
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: One day I want to figure out how this happened.

At least for mingw it somewhat makes sense - the win64 ABI does something unusual by special-casing only powers of 2 (and 0 isn't a power of 2), and the interaction with the GNU extension allowing ZSTs was likely accidentally overlooked.

But what's up with the other architectures? Why do they care about ZSTs at all? Were these all bugs in the GCC support?

File: src/librustc/ty/layout.rs [hunk 2]
```diff
@@ -2723,11 +2722,10 @@ where
...
             let mut arg = mk_arg_type(ty, arg_idx);
             if arg.layout.is_zst() {
-                // For some forsaken reason, x86_64-pc-windows-gnu
-                // doesn't ignore zero-sized struct arguments.
-                // The same is true for s390x-unknown-linux-gnu
-                // and sparc64-unknown-linux-gnu.
-                if is_return || rust_abi || (!win_x64_gnu && !linux_s390x && !linux_sparc64) {
+                // FIXME: The C ABI case should be handled in adjust_for_cabi.
+                // Zero-sized struct arguments cannot be ignored in the C ABI
+                // if they are passed indirectly.
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: This makes it sound like it's normal to a pass a pointer to nothing, and not likely some historical accident :disappointed:.

(To be clear, I'm not asking for a change here, my wording was probably too strong anyway)

File: src/librustc/ty/layout.rs (hunk 1 above)
Comment (smaeul) [reply src/librustc/ty/layout.rs]: @eddyb The spec and the code driving this are quite simple: aggregates are passed by reference, full stop. There is no exception based on the size of the aggregate. Or in other words, they *don't* care about ZSTs, while rust does.

The PowerPC SVR4 psABI is linked above, and the gcc code driving this is [rs6000_pass_by_reference()](http://gcc.gnu.org/git/?p=gcc.git;a=blob;f=gcc/config/rs6000/rs6000.c;h=cb8d737976b24f52115b8fb316b74337e0100530;hb=3e7b85061947bdc7c7465743ba90734566860821#l12224) (with the `if (DEFAULT_ABI == ABI_V4 && AGGREGATE_TYPE_P (type))` condition on line 12254).

File: src/librustc/ty/layout.rs (hunk 2 above)
Comment (smaeul) [reply src/librustc/ty/layout.rs]: The contents of the pointer don't really matter (the callee will never dereference it, although C++ guarantees that if you take the address of a ZST it will be unique). The important part is that we reserve a register or stack slot for the pointer, since the spec guarantees that there will be a pointer provided. Otherwise (and this was what was causing test failures), all arguments after the ZST will be misaligned from what the C compiler expects.

File: src/librustc/ty/layout.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc/ty/layout.rs]: > Or in other words, they *don't* care about ZSTs, while rust does.

AFAIK they don't mention ZSTs because ZSTs are not in the C standard, so the GCC implementation is a defacto standard for the ZST GNU extension.

What I think happened was that:
* GCC added this ZST extension
* GCC *didn't* bypass target-specific logic for ZSTs (whereas we do by default)
* new targets were added to GCC *without* anyone involved bringing up the ZST extension

Here are a couple architectures: https://godbolt.org/z/yOCJ-z - you can change the compilers involved to other architectures to check what they do, in that example MIPS64 is correct while MSP430 has the bug.

We might even support MSP430 without being aware of the bug - this is becoming quite sad, IMO.

File: src/librustc/ty/layout.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc/ty/layout.rs]: The spec doesn't apply to ZSTs because the C standard bans ZSTs (see my other comment for more details).
I don't think we need to read intent into what is pretty clearly a historical accident, given that x64/ARM/MIPS all do the right thing while some less-common targets don't.

File: src/librustc/ty/layout.rs [hunk 3]
```diff
@@ -2648,12 +2648,11 @@ where
...
 
         let target = &cx.tcx().sess.target.target;
-        let win_x64_gnu =
-            target.target_os == "windows" && target.arch == "x86_64" && target.target_env == "gnu";
-        let linux_s390x =
-            target.target_os == "linux" && target.arch == "s390x" && target.target_env == "gnu";
-        let linux_sparc64 =
-            target.target_os == "linux" && target.arch == "sparc64" && target.target_env == "gnu";
+        let indirect_zst = match target.arch.as_ref() {
+            "powerpc" | "s390x" | "sparc64" => true,
```
Comment (eddyb) [inline src/librustc/ty/layout.rs]: Wait, you've removed the `target_env == "gnu"` - this workaround *only* makes sense in that case, because ZSTs in C are a GNU extension.

File: src/librustc/ty/layout.rs (hunk 1 above)
Comment (smaeul) [reply src/librustc/ty/layout.rs]: That's because aggregates are passed in registers on MIPS/MIPS64/PowerPC64 (so the ZST takes up zero registers), while aggregates are passed indirectly on PowerPC/MSP (so the pointer to the ZST takes up one register, just like any other pointer). You need a caller to see the full picture.: https://godbolt.org/z/iMnRe-

File: src/librustc/ty/layout.rs (hunk 3 above)
Comment (smaeul) [reply src/librustc/ty/layout.rs]: @eddyb That's not true. It makes equally as much sense for `target_env == "musl"`, which follows the exact same ELF ABI as glibc. As I mentioned before:
> With these patches, I get a clean testsuite run on `powerpc-unknown-linux-musl`.
It is also necessary for the BSDs, since again they use the same ABI (both as specified, and as implemented in gcc/clang).

Comment (eddyb) [reply src/librustc/ty/layout.rs]: You could probably add a `has_gnu_extensions` or something that's set for `target_env` being either `"gnu"` or `"musl"`.

While this doesn't feel ideal, I'd certainly prefer to broaden the whitelist than start including combinations we might not be aware of, and it does make sense that `musl` would hit the same problems.

**Comment by JohnCSimon** [comment]
Ping from triage.
@nagisa This PR has sat idle for the last few days. Can you please address the merge conflict? 
CC @eddyb 
Thank you!

**Comment by smaeul** [comment]
I've rebased this on `master`.

**Comment by JohnCSimon** [comment]
Ping from triage. This seems to be waiting on the review
@eddyb can you please review?

cc: @smaeul 

Thanks!

**Comment by ProgrammaticNajel** [comment]
Ping from triage. @eddyb any updates on this? Thanks.

**Comment by eddyb** [comment]
This PR has been waiting on this comment, not me: https://github.com/rust-lang/rust/pull/64259#discussion_r329459031.

**Comment by joelpalmer** [comment]
Ping from Triage:  Closing due to inactivity. @smaeul please reopen with any updates! Thanks for the PR.

File: src/librustc/ty/layout.rs (hunk 3 above)
Comment (smaeul) [reply src/librustc/ty/layout.rs]: That still doesn't make any sense to me. `target_env` refers to the libc in use: glibc versus musl. That has nothing to do with the ABI, which is a property of the compiler.

Like I already said, this fix applies for the BSDs as well as Linux (regardless of `target_env`), because they all use GCC or a GCC-compatible compiler.

**Comment by smaeul** [comment]
@joelpalmer @eddyb Please reopen this. I haven't made any changes because I believe the code is correct as-is.

**Comment by eddyb** [comment]
I've reopened this and nominated it for @rust-lang/compiler team discussion.

<hr/>

I still disagree that we should expand the ZST whitelist without being fully precise about the problematic targets, and I maintain that they are a byproduct of the GNU extension allowing ZSTs (when the C standard does not, and therefore the psABI does not need to specify them), and GCC implementations of those targets overlooking that extension, with GCC not having an unified ZST check like we do.

I would be interested in historical messages on the GCC mailing list that show otherwise.

<hr/>

Alternatively, we could claim that we don't support the GNU ZST extension to C (and lint against ZSTs in FFI), but I'm not sure nobody uses that (cc @joshtriplett @gnzlbg).

**Comment by joshtriplett** [comment]
@eddyb Projects do use that occasionally, yes, so the ability to interoperate with it is important. We would in general need to match GCC's ABI for that. If GCC's ABI for that is *broken* on a target,we should find out from the GCC folks whether that's considered a bug in GCC that should be fixed or whether that's just the behavior going forward now for compatibility.

**Comment by eddyb** [comment]
The behavior has been around for long enough that I doubt changing it is possible.

I think the first target we've observed this on was mingw, btw - likely due to Microsoft hardcoding pow2 sizes in an ABI, i.e. `[u8; 3]` and `()` are both passed indirectly, while `[u8; 4]` is passed like `u32`.
At least there we might be able to say that we don't support by-value ZSTs, as they don't make sense with MSVC, but it's trickier with platforms where GCC or (GCC-compatible) Clang are the main compiler.

Note that this is about passing ZSTs *by value* as arguments, or returning them, *not* using them behind pointers (which works fine regardless of what we do to call ABI).

**Comment by eddyb** [comment]
AFAIK x86, x86_64, ARM, AArch64, MIPS and MIPS64 all ignore ZSTs correctly, see https://github.com/rust-lang/rust/pull/64259#discussion_r329124633.

**Comment by lu-zero** [comment]
@Edelsohn could you help us?

**Comment by parched** [comment]
> AFAIK x86, x86_64, ARM, AArch64, MIPS and MIPS64 all ignore by-value ZSTs correctly.

As far as I can remember that's not correct for AArch64 https://github.com/rust-lang/rust/issues/58487.

I'll ping the folks at Arm who write the ABI, but I suspect it will be unlikely to be able from anything GCC does.

**Comment by edelsohn** [comment]
"PowerPC" is 32 bit PowerPC ABI (ABI_V4).  The PPC64LE Linux ABI is referenced as ABI_ELFv2.  The PPC64BE ABI is ABI_AIX because it mostly matches the 64 bit AIX ABI.  The parameter passing is not by reference for those prominent ABIs.

IIUC there is no problem for PPC64 Linux ABIs.

**Comment by eddyb** [comment]
> As far as I can remember that's not correct for AArch64 #58487.

@parched #58487 sounds like it's about ZSTs in non-ZST types, whereas here we're talking about passing *an entire ZST* into or out of a function, as a whole argument/return value.

> IIUC there is no problem for PPC64 Linux ABIs.

@edelsohn according to other comments on this issue, that seems to be correct.

**Comment by smaeul** [comment]
@eddyb Of course you're only going to see indirectly-passed by-value ZSTs on a subset of architectures, because only a subset of architectures have *any* indirectly-passed by-value struct arguments. This situation simply doesn't apply to ABIs where by-value structs are always passed in registers. That doesn't in any way mean some ABIs are "correct", or some are "incorrect", or that GCC has a bug. The ABI is a convention, an *existing* convention, and it's largely arbitrarily chosen. If `rustc` isn't following it, then that's a `rustc` bug, not a bug in all of the existing software following the ABI.

**Comment by gnzlbg** [comment]
@smaeul the issue here is that "all of the existing software following the ABI" is probably zero when it comes to ZST fields in non-ZST structs because C is the lingua franca for ABI interoperation and portable C code does not have ZST fields in non-ZST structs. Whether the ABI specs should be special cased to improve this situation is an issue worth raising with the ABI maintainers themselves, even if the decision ends up just being "just keep doing what GCC does just in case somebody relies on that".

**Comment by smaeul** [comment]
@gnzlbg That sounds reasonable. It also sounds like it will take a while. Should I open a separate PR for the first commit, which fixes the ABI for non-ZST structs? It seems to be uncontroversial. That way we can go ahead and fix the common case while figuring out what to do in this special case.

**Comment by eddyb** [comment]
> Should I open a separate PR for the first commit, which fixes the ABI for non-ZST structs?

Yes please! I'm really sorry I didn't think of this earlier, r=me on the first commit.

File: src/librustc_target/abi/call/powerpc.rs [hunk 4]
```diff
@@ -1,49 +1,28 @@
...
+fn classify_arg_ty<Ty>(arg: &mut ArgType<'_, Ty>) {
     if arg.layout.is_aggregate() {
-        arg.cast_to(Uniform {
-            unit: Reg::i32(),
-            total: size
-        });
-        if !offset.is_aligned(align) {
-            arg.pad_with(Reg::i32());
-        }
+        arg.make_indirect();
```
Comment (eddyb) [inline src/librustc_target/abi/call/powerpc.rs]: So I went ahead and checked all other ABIs that use `make_indirect` on arguments, and *all* of them, with no exception, check the size first before calling `make_indirect`, so this ABI is unique (assuming the others are even correct, that is).

**Comment by edelsohn** [comment]
PPC64BE (AIX ABI) and PPC64LE (ELFv2 ABI) do not consume a register or memory for ZST.  This seems to be the preferred behavior and what the Rust community desires.

The PPC32 SVR4 ABI designed by Sun passes structs by reference and consumes a register for the pointer to a ZST.

PPC64BE Linux distributions deprecated 32 bit support.  PPC64LE Linux distributions do not support 32 bit at all.  PPC32 ELF primarily is used for the embedded market.  Any new support for PPC32 in Linux distributions targeted at the server market likely would be a variant of ELFv2 ABI.  (32 bit AIX uses the AIX ABI, which has correct behavior for ZST. Darwin also uses an ABI related to AIX.)

The "SVR4" ABI doesn't specify the behavior for ZST. The consumption of a register seems to be consistent with the rest of the ABI although inefficient.  The ownership / copyright for the SVR4 ABI is unclear.  There is not much focus on PPC32 Linux/ELF from the major Power/PowerPC vendors, so I don't believe that anyone is going to champion documenting the behavior in the ABI or fixing the ABI. 

The Rust community and the PowerPC community already have devoted more time to the ZST issue on PPC32 V4 ABI than it probably deserves.  In terms of chip volumes, there are a huge number of embedded 32 bit PowerPC customers (relative to server chip volume) who probably have frozen their development environment and unlikely ever will use Rust.

**Comment by pnkfelix** [comment]
We discussed this a bit in the T-compiler meeting: [started here](https://zulip-archive.rust-lang.org/131828tcompiler/04503weeklymeeting2019110754818.html#180145831) and continued [here](https://zulip-archive.rust-lang.org/131828tcompiler/69822PowerPCCABIfixesPR64259.html) after I migrated to a dedicated topic.

**Comment by glaubitz** [comment]
@pnkfelix What @gnzlbg says is not correct: The ppc32 is being used in Debian, in openSUSE and I think there is also a Void Linux port for ppc32 which uses Rust, see: https://github.com/void-ppc/void-packages/commit/637c28ccc35e2ceb7471c071b71a8b02b0d572dc

CC @jrtc27 @q66

**Comment by q66** [comment]
Yes, that is correct. We've been shipping these patches downstream since 1.38, and Firefox no longer segfaults rustc :) (but it doesn't start either way)

**Comment by glaubitz** [comment]
Do you also get lower testsuite failures? I haven't fully tested the patches yet.

**Comment by q66** [comment]
It should, though currently building 1.39 with it. Also, what @edelsohn is saying is incorrect; the 32-bit PowerPC ABI is unrelated to what the 64-bit stuff is using (ELFv2/ELFv1). The Void 64-bit LE and BE ports are both ELFv2 ABI (ELFv2 is not tied to any endianness) with the BE port being able to run 32-bit binaries (using the single 32-bit PowerPC ABI, same as it has always been)

**Comment by edelsohn** [comment]
@q66 I never stated that 32 bit SVR4 PowerPC ABI is related to ELFv1/ELFv2 ABI. I referred to AIX PPC32 ABI, not Linux.  The AIX 32 bit ABI is the same as the AIX 64 bit ABI which is similar to the ELFv1 ABI. PPC64BE systems use ELFv1 ABI for 64 bit and SVR4 ABI for 32 bit.

Please quote the passage that you believe is wrong or withdraw your claims about what I stated.

**Comment by q66** [comment]
for example, just now: "PPC64BE Linux systems use ELFv1 ABI for 64 bit and SVR4 ABI for 32 bit." which you also said in your previous post in a way.

PPC64BE systems can use either ELFv1 or ELFv2 ABI. There are multiple distributions at this point that do so, the musl libc does so exclusively, glibc also works fine (we're shipping it).

"PPC64BE Linux distributions deprecated 32 bit support." is also wrong, according to @glaubitz it continues in Debian, we're shipping it in Void (not deprecated) and Adélie ships it as well (not deprecated).

**Comment by edelsohn** [comment]
@q66 Red Hat, SUSE and, I believe, Ubuntu all removed PPC32 installation from their PPC64BE distribution by default.  Earlier the distributions had been built as PPC32 default, even on 64 bit systems (PPC64 kernel).  They switched to building all libraries and applications as PPC64 while continuing to install PPC32 runtime.  Later they removed default installation of 32 bit environment, although still available as an optional install.

There also is one distribution that is using ELFv2 in BE mode.  The code is FOSS, so people can mix and match any features that they want and create a distribution.  The majority of users are not running with these unusual configurations and these unusual configuration are not a priority for commercial support or most kernel and toolchain maintainers for Power.

You seem to be using the word "using" in two different ways: "the 32-bit PowerPC ABI is unrelated to what the 64-bit stuff is using (ELFv2/ELFv1)".  The ABIs are not related. I never stated that they were related.  Red Hat, SUSE, Ubuntu, Debian and Alpine have not shipped with PPC32 based on ELFv1 or ELFv2. Red Hat, SUSE and Ubuntu "use" the PPC32 SVR4 ABI with PPC64BE ELFv1 ABI for PPC64BE distros.

What is "64-bit stuff"?

You claimed that I was wrong and then confirmed my statements.

**Comment by glaubitz** [comment]
I work for SUSE, we are still building ppc32 for openSUSE.

Can we get those fixes merged? I think those discussions are not helping the cause.

**Comment by q66** [comment]
As I said above, there are multiple distributions using ELFv2 in BE mode, not just one.

Red Hat and Ubuntu deprecated even their PPC64BE support so they are irrelevant in this context. I have no idea about SUSE.

It doesn't matter what majority of users run or what has commercial support. The point is several distributions that use ELFv2 BE exist, and they all have 32-bit support out of box as well (PPC32 SVR4 ABI, either standalone or as userland compatibility), and are not deprecated. It doesn't matter what IBM chooses to support either, at least in this context.

To summarize:

1) Void, Adélie and Gentoo support 64-bit BE based on ELFv2; they also support 32-bit BE based on SVR4 ABI (either as a standalone 32-bit distribution, or as a part of the 64-bit distribution)
2) Debian and probably SUSE support 64-bit BE based on ELFv1, with 32-bit based on SVR4 ABI
3) Neither of those is deprecated

**Comment by glaubitz** [comment]
I have never heard that a distribution would use ELFv2 for ppc64 BE. Can you name any?

For all distributions that I know which have a ppc64 BE port had this port before ELFv2 came around, so I would be very surprised that any distribution using the new ABI exists.

ELFv2 was introduced with the ppc64el ports and I don't think any distribution exists that would intentionally break their ABI compatibility to switch to ELFv2 when users would just switch to POWER8 or newer anyway. And as far as I know, there 32-Bit PowerPC little-endian doesn't exist in the wild, if it works at all.

I'm therefore not sure I understand the argument why this change would break anything.

**Comment by q66** [comment]
@glaubitz I just named two binary-based ones in the post above....

In Void we have both LE and BE (people can choose whichever they want; the BE port runs on old and new hardware, the LE port only on new hardware), both based on ELFv2 ABI (it's a new port, so legacy does not matter). Adélie is BE only, based on ELFv2 ABI. The latter is based on the musl libc, which only supports ELFv2. The former can use either glibc or musl, both with ELFv2.

**Comment by glaubitz** [comment]
Ok, I wasn't aware that ELFv2 on ppc64be actually existed in the wild :). And in particular, I didn't know musl is ELFv2 only, so thanks for the explanation.

But, anyway, the 32-bit PowerPC ABI has some issues and it seems these patches fix those issues, so I'm not sure why these patches shouldn't get merged. We have fixed other ABI issues in Rust before, for example in the sparc64 port.

**Comment by q66** [comment]
I agree completely, and the discussion above is unrelated to that, so it would be best to end it there in order to not clutter the pull request.

If you decide to downstream those patches, please do keep in mind that they're breaking as far as codegen plugins go, so you can *not* bootstrap a new Rust installation natively using a broken toolchain (this is because Rust uses those plugins between stages, and mixing is no good). So you will have to cross-compile a new bootstrap toolchain (cross-compiling works fine, there is no chance of plugins leakage for that) and use that to bootstrap a fresh set on the target system.

**Comment by glaubitz** [comment]
I think I actually ran into this problem and I understand what went wrong. This also explains the reluctance on the Rust side regarding this change.

**Comment by q66** [comment]
It shouldn't be problematic for Rust itself, I'm fairly sure they cross compile all of their tier 2+ targets anyway - so once it's merged and a release is made with it, all of the Mozilla-released tarballs should be suitable for bootstrapping from scratch...

**Comment by awilfox** [comment]
Adélie does have a PPC32 port which uses the SVR4 ABI, with an install base (at least 50 at my count), and most of these people do enjoy having Firefox, which builds with Rust, which means Rust needs to be fixed.

To clarify, yes, we also have a PPC64BE port.  This uses ELFv2, and works on 970 and up (G5, P5+).  This has an even higher install base (a few hundred at my count).  It's used in production for virtually all adelielinux.org services.  We've worked with major projects across the ecosystem to ensure ELFv1/ELFv2 is a separate concern from BE/LE.  Upstreams use auxv to determine VSX and VMX support so the silly requirement that ELFv2 is "only for POWER8" is irrelevant.  Especially with things like [microwatt](https://github.com/antonblanchard/microwatt) and the opening of OpenPOWER, the ELFv2 ABI likely should be changed to note that auxv should be used instead of assuming an ISA level.  But that's all irrelevant, I guess.

Please do merge this upstream so we can stop carrying it.  It is still useful, whether or not chips are "being produced".  Wii/Wii U, Freescale chips, Apple G3/G4, there are *millions* of devices right now that use ppc32.  See https://github.com/ibmruntimes/node/issues/30 for another instance of ppc32 being neglected despite users desiring support.

**Comment by glaubitz** [comment]
@awilfox Thanks for your comment. I agree with everything what you said and I highly appreciate the work you guys are doing on the PowerPC code. We profit from these fixes in openSUSE and Debian as well!

**Comment by JohnCSimon** [comment]
Ping from triage: 
@smaeul  - Can you please address the merge conflict?
cc: @eddyb 

Thanks!

**Comment by JohnCSimon** [comment]
Ping from triage 
@eddyb can you please review this? Thank you

**Comment by Dylan-DPC** [comment]
This is being discussed on zulip so there is progress

**Comment by pnkfelix** [comment]
> This is being discussed on zulip so there is progress

To be honest I'm not aware of discussion beyond that shown in https://zulip-archive.rust-lang.org/131828tcompiler/69822PowerPCCABIfixesPR64259.html, where (as of now) the most recent comment was posted on November 7th.

@Dylan-DPC did discussion continue after that in some other Zulip topic?

(It would be good to try to get unblocked here...)

**Comment by Dylan-DPC** [comment]
Ah I probably misread it thinking it's latest. Sorry for that

**Comment by pnkfelix** [comment]
swiping from @eddyb to move this forward

**Comment by glaubitz** [comment]
Is there actually stuff left to merge? The ABI fixes that were shipped with 1.40 actually reduced the number of testsuite failures for us on Debian powerpc dramatically.

We're now down to only 23 failures, from over 60.

**Comment by nikomatsakis** [comment]
We discussed this in our compiler team triage meeting today. We'd very much like to get *some* version of these changes landed. @eddyb suggested there is a small (approx. 1 line) diff that would fix the immediate/known problems that they'd prefer, and there was general consensus that a small, conservative step would be a good place to start. Hopefully @eddyb can either comment here (and we can update this PR) or it may be easier for them to just open a PR with the diff.

@smaeul, I do wish to thank you both for the initial PR and for your patience!

**Comment by pnkfelix** [comment]
Discussed in T-compiler meeting. Closing under assumption that remaining desired work will be covered by PR's like #69263

**Comment by glaubitz** [comment]
This has reduced the number of testsuite failures on Debian/powerpc down to 10.

On the other side, it's currently not possible to bootstrap rustc on openSUSE using the binaries provided by the Rust website. It seems that the Rust binaries for powerpc are still built with a toolchain without these fixes, isn't it?

**Comment by mati865** [comment]
@glaubitz https://github.com/rust-lang/rust/pull/69263 is the beta but hasn't reached stable yet.

**Comment by glaubitz** [comment]
@mati865 I'm not sure whether that's related, is it?

And more surprisingly, the number of testsuite failures went up to 51 with 1.42.0 again.

**Comment by mati865** [comment]
@glaubitz this PR has been closed under assumption #69263 is enough to fix PPC issue.
